File: internal/paths/walk.go
// Package paths finds the files a run may act on and narrows them per task.
package paths

import (
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"

	ignore "github.com/sabhiram/go-gitignore"
)

// gitDir is never walked. It is not source, and nothing in a jig can want it.
const gitDir = ".git"

// A Matcher decides whether a path is ignored. Reading `.gitignore` as text is
// the whole of it: bolt invokes no git, so a run over a tree that is not a
// repository behaves exactly as one over a repository.
//
// It is an interface because the pattern matching behind it is the one part of
// the walk borrowed rather than written, and swapping it should be one file.
type Matcher interface {
	Ignored(relative string, isDir bool) bool
}

// Walk returns every file under base that a task may act on, as paths relative
// to base, in sorted order.
//
// Sorted because FR-9.4's identical work directory names across two runs rest
// on the matched list being the same list, and nothing else gives that.
//
// exclude names directories the walk must not descend into whatever
// `.gitignore` says. A run's own output directory goes here, because a tree
// accumulating runs would otherwise feed each run the last one's evidence.
func Walk(base string, exclude []string) ([]string, error) {
	excluded := make(map[string]bool, len(exclude))
	for _, dir := range exclude {
		if dir != "" {
			excluded[filepath.Clean(dir)] = true
		}
	}

	stack := newIgnoreStack(base)
	var found []string

	err := filepath.WalkDir(base, func(path string, entry fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		relative, relErr := filepath.Rel(base, path)
		if relErr != nil {
			return relErr
		}
		if relative == "." {
			return nil
		}

		if entry.IsDir() {
			if entry.Name() == gitDir || excluded[relative] {
				return filepath.SkipDir
			}
			if stack.ignored(relative, true) {
				return filepath.SkipDir
			}
			stack.enter(relative)
			return nil
		}

		// A symlink is not followed. Following one leaves the base and breaks
		// containment, and shared jigs arrive as tracked symlinks, so this is
		// the ordinary case rather than a hostile one.
		if entry.Type()&os.ModeSymlink != 0 {
			return nil
		}
		if !entry.Type().IsRegular() {
			return nil
		}
		if stack.ignored(relative, false) {
			return nil
		}

		found = append(found, filepath.ToSlash(relative))
		return nil
	})
	if err != nil {
		return nil, err
	}

	sort.Strings(found)
	return found, nil
}

// ignoreStack applies each `.gitignore` to the subtree it sits in, which is
// what git does and what a single root-level read would get wrong for any
// project keeping one further down.
type ignoreStack struct {
	base  string
	rules map[string]*ignore.GitIgnore
}

func newIgnoreStack(base string) *ignoreStack {
	stack := &ignoreStack{base: base, rules: map[string]*ignore.GitIgnore{}}
	stack.enter(".")
	return stack
}

// enter reads the `.gitignore` in dir, if there is one, so it applies from
// here down.
func (s *ignoreStack) enter(dir string) {
	path := filepath.Join(s.base, dir, ".gitignore")
	compiled, err := ignore.CompileIgnoreFile(path)
	if err != nil {
		// No `.gitignore` here, or an unreadable one. Neither is a reason to
		// refuse a run: the tree may not be a repository at all.
		return
	}
	s.rules[filepath.Clean(dir)] = compiled
}

func (s *ignoreStack) ignored(relative string, isDir bool) bool {
	subject := filepath.ToSlash(relative)
	if isDir {
		subject += "/"
	}

	for dir, rules := range s.rules {
		within, ok := under(dir, relative)
		if !ok {
			continue
		}
		candidate := filepath.ToSlash(within)
		if isDir {
			candidate += "/"
		}
		if rules.MatchesPath(candidate) {
			return true
		}
	}
	return false
}

// under says whether relative sits inside dir, and what it is called from
// there, which is the form a `.gitignore` in dir matches against.
func under(dir, relative string) (string, bool) {
	if dir == "." {
		return relative, true
	}
	prefix := dir + string(filepath.Separator)
	if !strings.HasPrefix(relative, prefix) {
		return "", false
	}
	return strings.TrimPrefix(relative, prefix), true
}

File: src/walk.rs
//! Finding the files a run's tasks act on.
//!
//! FR-2.2 has bolt walk the directory it was given. There is no
//! changed-since-a-ref: the walk is the whole input.

use std::path::{Path, PathBuf};

use crate::Error;

/// Walk `base`, returning the paths a task may act on.
///
/// The walk honours `.gitignore` by FR-2.2a, which keeps `.git`, `node_modules`,
/// a virtualenv and build output out of every run without a second list to
/// maintain.
///
/// FR-2.2b bounds how that is done. Honouring `.gitignore` means reading those
/// files as text: bolt does not invoke git, read anything under `.git/`, or
/// require a repository. The `ignore` crate reads git's own excludes at its
/// defaults, so each is turned off against the row rather than taken as it
/// comes, and `parents` with them, because a `.gitignore` above the base is a
/// file outside it changing the run's input, which FR-2.3 forbids.
///
/// Paths come back sorted, by FR-2.2d, which is what makes the matched list the
/// same list on every run over the same tree. FR-9.4's identical work directory
/// names rest on this and on nothing else.
///
/// # Errors
///
/// [`Error::BaseMissing`] when `base` is not there, by FR-2.5.
pub fn walk(base: &Path) -> Result<Vec<PathBuf>, Error> {
    if !base.is_dir() {
        return Err(Error::BaseMissing(base.to_path_buf()));
    }

    let mut found: Vec<PathBuf> = ignore::WalkBuilder::new(base)
        // FR-2.2b, each one turning off a default that reaches outside the row.
        .require_git(false)
        .git_global(false)
        .git_exclude(false)
        .parents(false)
        // FR-2.2e. Following one leaves the base and breaks FR-2.3's
        // containment. This is the crate's default and is set anyway, because
        // the row is the reason rather than the default being convenient.
        .follow_links(false)
        .build()
        .filter_map(Result::ok)
        .filter(|entry| {
            // Files, not directories: FR-2.2 is about the files tasks act on.
            //
            // A symlink is excluded here too, and that is PROVISIONAL. Measured
            // 2026-08-27: with `follow_links` off, `ignore` yields the link
            // itself, so a task handed one reads through it to outside the base.
            // FR-2.2e forbids following and does not say whether returning is
            // also forbidden, which is question 38's neighbour, question 40.
            // Excluding is the containment-preserving direction and is reversed
            // by deleting this predicate once the row exists.
            entry.file_type().is_some_and(|kind| kind.is_file())
        })
        .map(ignore::DirEntry::into_path)
        .collect();

    found.sort();
    Ok(found)
}
